From 5a04dc554ef53ddc39895a9211286b9fb2e65586 Mon Sep 17 00:00:00 2001
From: Maja Kabiljo <majakabiljo@fb.com>
Date: Tue, 3 Nov 2015 11:01:22 -0800
Subject: [PATCH] GIRAPH-1039: Fix stopping jmap histo thread

Summary: Currently if jmap histo frequency is set to long period we end up stuck in the end of the job for a long time waiting on jmap histo thread

Test Plan: Ran a job with long jmap frequency - verified it gets stuck without this change and finishes fine with it

Differential Revision: https://reviews.facebook.net/D50085
---
 .../main/java/org/apache/giraph/utils/JMapHistoDumper.java   | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/giraph-core/src/main/java/org/apache/giraph/utils/JMapHistoDumper.java b/giraph-core/src/main/java/org/apache/giraph/utils/JMapHistoDumper.java
index 4282d3503..a68f6c44b 100644
--- a/giraph-core/src/main/java/org/apache/giraph/utils/JMapHistoDumper.java
+++ b/giraph-core/src/main/java/org/apache/giraph/utils/JMapHistoDumper.java
@@ -42,7 +42,7 @@ public class JMapHistoDumper implements MasterObserver, WorkerObserver {
   /** The jmap printing thread */
   private Thread thread;
   /** Halt jmap thread */
-  private boolean stop = false;
+  private volatile boolean stop = false;
 
   @Override
   public void preLoad() {
@@ -70,6 +70,7 @@ public void postApplication() {
   private void joinJMapThread() {
     stop = true;
     try {
+      thread.interrupt();
       thread.join(sleepMillis + 5000);
     } catch (InterruptedException e) {
       LOG.error("Failed to join jmap thread");
@@ -89,7 +90,7 @@ public void run() {
           try {
             Thread.sleep(sleepMillis);
           } catch (InterruptedException e) {
-            LOG.warn("JMap histogram sleep interrupted", e);
+            LOG.info("JMap histogram sleep interrupted", e);
           }
         }
       }
